core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java

```diff
@@ -215,6 +223,85 @@
+        }
+        return Optional.of(repeatable.get().value());
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T extends Annotation, E> E getAnnotationField(T annotation, String fieldName,
+                                                                  Class<E> cast) {
+        return (E) getAnnotationField(annotation, fieldName);
+    }
+
+    private static <T extends Annotation> Object getAnnotationField(T annotation, String fieldName) {
+        try {
+            return annotation.getClass().getMethod(fieldName).invoke(annotation);
+        } catch (Exception ex) {
+            throw new IllegalStateException(
+                    String.format("'%s.%s()' method raised an exception.",
+                                  annotation.getClass().getName(), fieldName), ex);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T extends Annotation> List<T> findRepeatableAnnotations(AnnotatedElement element,
+                                                                            Class<T> annotationType) {
+        requireNonNull(element, "element");
+        requireNonNull(annotationType, "annotationType");
+        final Class<? extends Annotation> repeatable = getRepeatable(annotationType).orElseThrow(
+                () -> new IllegalArgumentException("annotationType is not a repeatable annotation."));
+        return getAllAnnotations(element).stream()
+                                         .filter(a -> a.annotationType() == annotationType ||
+                                                      a.annotationType() == repeatable)
+                                         .flatMap(annotation -> {
+                                             if (annotation.annotationType() == repeatable) {
+                                                 return Arrays.stream(
+                                                         (T[]) getAnnotationField(annotation, "value"));
+                                             }
+                                             return Stream.of((T) annotation);
+                                         }).collect(toImmutableList());
+    }
+
+    private static <T extends Annotation> void setAdditionalHeader(HttpHeaders headers,
+                                                                   AnnotatedElement element,
+                                                                   String clsAlias,
+                                                                   String elementAlias,
+                                                                   String level,
+                                                                   Class<T> annotation) {
+        requireNonNull(headers, "headers");
+        requireNonNull(element, "element");
+        requireNonNull(level, "level");
+
+        final Set<String> addedHeaderSets = new HashSet<>();
+        findRepeatableAnnotations(element, annotation).forEach(header -> {
```

Comment:
I'm not sure if it works. But would you try to apply `AnnotationUtil#findAll` for this?

https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotationUtil.java#L125